File: problem2-dfs.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""
from collections import deque
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        if not employees: return 0
        self.hashmap = {}
        self.total_importance = 0
        for emp in employees:
            self.hashmap[emp.id] = emp
        self.dfs(id)
        return self.total_importance
    
    def dfs(self, id):
        #base 
        if self.hashmap[id].subordinates == None:
            return
        #logic
        e = self.hashmap[id]
        self.total_importance += e.importance
        for sub in e.subordinates:
            self.dfs(sub)
            
        
            
#Time complexity is O(n) and space complexity is O(n)
#DFS approach used for evaluating all the subordiantes of the given id.
            

File: Problem2_DFS.java
//Time Complexity: O(n)
//Space Complexity: O(n)
//Code run successfully on LeetCode.

public class Problem2_DFS {

    HashMap<Integer, Employee> map;
    public int getImportance(List<Employee> employees, int id) {
        
        if(employees == null || employees.size() == 0)
            return -1;
        
        map = new HashMap<>();
        
        for(Employee e : employees){
            map.put(e.id, e);
        }
        
        return dfs(id);
    }
    
    private int dfs(int id){
        
        Employee e = map.get(id);
        
        if(e.subordinates == null)
            return e.importance;
        
        int total = 0;
        for(int i : e.subordinates){
             total += dfs(i);
        }
        
        return total + e.importance;
    }
}
